Make succeeded and failed in ICB final + introduce onSuccess

lorban opened the pull request:
Closes #11932

lorban committed:
#11932 make succeeded and failed in ICB final + introduce onSuccess

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#11932 refactor H3 message flusher to have ICB succeeded and failed final

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@gregw I went over all our ICB implementations that override `succeeded` and `failed`, and I came up with two conclusions:

 - Overriding `succeeded` is often used to release a successfully written buffer before moving on to the next one. But that can easily be replaced by a new `onSuccess` protected method that can do the same job but in a safe way.
 - Overriding `failed` is sometimes done to release a buffer, but that can be done in `onCompleteFailure` instead.
 - There are two places where `succeeded` and `failed` are overridden for different reasons, and IMHO those are pure evil. See `CustomTransportTest` and H3's `MessageFlusher` to see what I mean: the 1st one calls `iterate()` from `succeeded` and does not call super, the 2nd one calls `super.succeeded()` from `failed()`.

I'm now looking for a way to modify `MessageFlusher` so that it doesn't need to act this weird, but I'm building a firm belief that overriding `succeeded` and `failed` is never, ever a good idea.

lorban commented:
I refactored `MessageFlusher` and make ICB's `succeeded` and `failed` methods final.

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:
```diff
@@ -167,6 +167,13 @@ protected IteratingCallback(boolean needReset)
     protected abstract Action process() throws Throwable;
 
+    /**
+     * Invoked when one task has completed successfully.
```
More javadoc required about how this is serialized etc.
Specifically that it will not be called if the ICB is aborted and then succeeded.

lorban replied:
Ok, I improved javadoc.

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:
```diff
@@ -342,7 +356,7 @@ else if (notifyCompleteFailure != null)
      */
     @Override
-    public void succeeded()
+    public final void succeeded()
```
We can't make these `final` in 12.0.x.   But I do think this PR should be developed with them final so we can potentially make them final in 12.1.x.  We can remove the actual "final" just before merge.

Can you modify the javadoc to say that these methods should be treated as essentially final and only overridden in extraordinary circumstances.  I know nobody reads javadoc, but at least we can point to that if/when we ultimately do make them final.

lorban replied:
Agreed, once we're happy with this PR, I'll revert these `final` markers as the last step before merging.

gregw replied:
Remove those now and I'll give the green tick

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java:
I'm somewhat concerned by the current implementation of ICB and thought we made a lot of good improvements in #11876 to the implementation.

Can you look at the additional tests done in #11876 about testing abort in all states and do something similar in this PR.    If the current implementation passes, then good... else we should consider taking the new implementation (but with tuned down semantics).

lorban replied:
I'm not sure it's worth it, I even think I should rebase this PR on top of 12.1.x as we're heading into the direction of removing buffers from the pool for 12.0 and not implementing abortable callbacks.

So all changes to `IteratingCallback` should retarget 12.1.x and 12.0.x should be left as-is IMHO. This would also avoid making this an "omnibus" PR, as we've done a bit too much recently.

gregw replied:
If this PR and the one on the remove buffers fix the issues, then yes, let's leave further ICB refinement and testing to 12.1

gregw commented on jetty-core/jetty-quic/jetty-quic-common/src/main/java/org/eclipse/jetty/quic/common/QuicConnection.java:
```diff
@@ -372,17 +372,9 @@ protected Action process()
 
         @Override
-        public void succeeded()
+        protected void onSuccess()
```
I'm wondering why we protect the `entry` field with a lock in `process` but not here.   Both methods are called from the same serialized processing, so both should have same locking and memory barriers.

gregw replied:
Ah is it that the lock is protecting the queue and not the entry field?  If so, some comments would be good.... or just use a concurrent queue.

lorban replied:
Good catch, I'm going to replace the queue with a thread-safe impl.

gregw commented on jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/internal/HTTP2Flusher.java:
```diff
@@ -294,7 +294,7 @@ protected Action process() throws Throwable
 
     @Override
-    public void succeeded()
+    protected void onSuccess()
```
I think this change alone might fix a lot of strangeness we get with failures!

lorban replied:
:+1:

gregw reviewed (changes requested):
Nice big step in the right direction!  But a few things....

lorban committed:
#11932 simplify queue handling

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#11932 improve javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@olamy do you have an idea why the integration tests are failing with a compilation error despite the fact that the code doesn't match what's reported?

olamy committed:
force jetty-client in invoker it tests

Signed-off-by: Olivier Lamy <olamy@apache.org>

olamy committed:
force jetty-client in invoker it tests

Signed-off-by: Olivier Lamy <olamy@apache.org>

olamy committed:
force jetty-client in invoker it tests

Signed-off-by: Olivier Lamy <olamy@apache.org>

lorban committed:
#11932 revert final modifier

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#11932 fix bug in case of invalid action + add test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@gregw I reverted the `final` modifiers in ICB, and I also fixed a small bug in the delayed `onSuccess` call and added a test.

I think this can now be merged.

lorban commented:
@olamy thanks for your help fixing the build!
